crates/autopilot/src/limit_orders/quoter.rs

```diff
@@ -33,40 +31,43 @@ pub struct LimitOrderQuoter {
 
 impl LimitOrderQuoter {
     pub fn spawn(self) {
-        tokio::spawn(async move {
-            loop {
-                if let Err(err) = self.update().await {
-                    tracing::error!(?err, "failed to update limit order surplus");
-                }
-                tokio::time::sleep(self.loop_delay).await;
-            }
-        });
+        tokio::spawn(async move { self.background_task().await });
     }
 
-    async fn update(&self) -> Result<()> {
+    async fn background_task(&self) -> ! {
         loop {
-            let orders = self
-                .database
-                .limit_orders_with_outdated_fees(self.limit_order_age)
-                .await?;
-            if orders.is_empty() {
-                break;
-            }
-            for chunk in orders.chunks(10) {
-                join_all(chunk.iter().map(|order| {
-                    async move {
-                        let quote = self.get_quote(order).await;
-                        self.update_fee(&order.metadata.uid, &quote).await;
-                    }
-                    .instrument(tracing::debug_span!(
-                        "surplus_fee",
-                        order =% order.metadata.uid
-                    ))
-                }))
-                .await;
-            }
+            let sleep = match self.update().await {
+                Ok(0) => Duration::from_secs_f32(10.),
+                Ok(_) => Duration::from_secs_f32(0.),
```

Comment:
Shouldn't this branch only happen if `orders.len() < PARALLELISM`. I don't think it makes much of a difference (since it will just trigger another DB lookup, and would return right away in case there are no new orders), but this surprised me a bit when trying to understand the code compared to the PR comment.

Maybe, it makes more sense for `update` to return a `bool` indicating whether or not there are "more updates to process". This would, IMO make the call here a little clear WRT the intention of each of the methods.

Very nitpicky though - not blocking merge from my point of view.

Reply:
`update` doesn't exactly know whether there are more updates to process. It knows that if it gets less than N orders, then at that moment there would be no more outdated orders. By the time `update` has completed there might now be outdated orders again.

If we have 0 outdated orders then we definitely want to sleep to prevent busy looping on the DB. If we had at least 1 order then `update` internally spent some time outside of the DB query so in that sense it has already slept.

Changing `Ok(0)` to `Ok(n) if N < PARALLELISM` would be reasonable too. Both versions look fine to me.

I also see the argument in favor of just returning a bool but then there is still the decision of which comparison should determine the bool. Like you saiud this is a small detail. I'll think about it more and make a decision.